EFCore.NamingConventions/Internal/NameRewritingConvention.cs

```diff
@@ -21,17 +26,60 @@ public class NameRewritingConvention :
         {
             var entityType = entityTypeBuilder.Metadata;
 
-            // Only touch root entities for now (TPH). Revisit for TPT/TPC.
-            if (entityType.BaseType == null)
+            // Note that the base type is null when the entity type is first added - a base type only gets added later
+            // (see ProcessEntityTypeBaseTypeChanged). But we still have this check for safety.
+            if (entityType.BaseType is null)
             {
                 entityTypeBuilder.ToTable(_namingNameRewriter.RewriteName(entityType.GetTableName()), entityType.GetSchema());
             }
         }
 
+        public void ProcessEntityTypeBaseTypeChanged(
+            IConventionEntityTypeBuilder entityTypeBuilder,
+            IConventionEntityType newBaseType,
+            IConventionEntityType oldBaseType,
+            IConventionContext<IConventionEntityType> context)
+        {
+            var entityType = entityTypeBuilder.Metadata;
+
+            if (newBaseType is null)
+            {
+                // The entity is getting removed from a hierarchy. Set the (rewritten) TableName.
+                entityTypeBuilder.ToTable(_namingNameRewriter.RewriteName(entityType.GetTableName()), entityType.GetSchema());
+            }
+            else
+            {
+                // The entity is getting a new base type (e.g. joining a hierarchy).
+                // If this is TPH, we remove the previously rewritten TableName (and non-rewritten Schema) which we set when the
+                // entity type was first added to the model (see ProcessEntityTypeAdded).
+                // If this is TPT, TableName and Schema are set explicitly, so the following will be ignored.
+                entityTypeBuilder.HasNoAnnotation(RelationalAnnotationNames.TableName);
+                entityTypeBuilder.HasNoAnnotation(RelationalAnnotationNames.Schema);
+            }
+        }
+
         public virtual void ProcessPropertyAdded(
-            IConventionPropertyBuilder propertyBuilder, IConventionContext<IConventionPropertyBuilder> context)
-            => propertyBuilder.HasColumnName(_namingNameRewriter.RewriteName(propertyBuilder.Metadata.GetColumnName()));
+            IConventionPropertyBuilder propertyBuilder,
+            IConventionContext<IConventionPropertyBuilder> context)
+        {
+            var entityType = propertyBuilder.Metadata.DeclaringEntityType;
+            var property = propertyBuilder.Metadata;
+
+            property.SetColumnName(_namingNameRewriter.RewriteName(property.GetColumnBaseName()));
 
+            foreach (var storeObjectType in _storeObjectTypes)
```

Comment:
Why do you rewrite column names for all identifier types, but only rewrite identifier name for tables?

Reply:
The logic here was that while table names are set by convention, view mapping is done by calling ToView which requires the view name to be set explicitly - is my mental model wrong? I understand that there can still be a convention (from a plugin, provider) which maps to views (is this what you had in mind?), so I've just added rewriting view names too. 

For DbFunction I guess the logic is the same - any situation apart from a plugin or provider convention where one would be created by convention?